codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java

```diff
@@ -526,32 +529,29 @@ private boolean exceedsMaxHeaderSize(long size) {
         return true;
     }
 
-    private byte[] readStringLiteral(InputStream in, int length) throws IOException {
-        byte[] buf = new byte[length];
-        if (in.read(buf) != length) {
-            throw DECOMPRESSION_EXCEPTION;
-        }
-
+    private CharSequence readStringLiteral(ByteBuf in, int length) throws IOException {
         if (huffmanEncoded) {
-            return Huffman.DECODER.decode(buf);
+            return new AsciiString(huffmanDecoder.decode(in, length), false);
         } else {
-            return buf;
+            byte[] buf = new byte[length];
+            in.readBytes(buf);
+            return new AsciiString(buf, false);
         }
     }
 
     // Unsigned Little Endian Base 128 Variable-Length Integer Encoding
-    private static int decodeULE128(InputStream in) throws IOException {
-        in.mark(5);
+    private static int decodeULE128(ByteBuf in) throws IOException {
+        in.markReaderIndex();
         int result = 0;
         int shift = 0;
         while (shift < 32) {
-            if (in.available() == 0) {
+            if (!in.isReadable()) {
                 // Buffer does not contain entire integer,
                 // reset reader index and return -1.
-                in.reset();
+                in.resetReaderIndex();
                 return -1;
             }
-            byte b = (byte) in.read();
+            byte b = in.readByte();
```

Comment:
can we use a `ByteProcessor` here? Seems like we could reduce some bounds checking.